### PR TITLE
Add support for StateFlow.

### LINKED ISSUE #726: `StateFlow` bridging support
`StateFlow`, in contrast to `Flow`, exposes a value synchronously which is critical for use in Compose UI as it means you do not need to introduce a dummy default value while you wait for a `Flow`'s first emission.

### PR DESCRIPTION
**Pull request by colinrtwhite**
Adds support for using `StateFlow` in a `ZiplineService`.

Resolves: https://github.com/cashapp/zipline/issues/726

### CODE REVIEW

**Comment by colinrtwhite** [comment]
Interesting - looks like one of the tests only fails on `ubuntu-latest`:

```
app.cash.zipline.StateFlowTest.collectFlowOneTime FAILED
    kotlin.AssertionError at null:-1
```

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/flows.kt [hunk 1]
```diff
@@ -86,3 +91,65 @@ internal class FlowSerializer<T>(
...
+
+  override fun deserialize(decoder: Decoder): StateFlow<T> {
+    val service = decoder.decodeSerializableValue(delegateSerializer)
+    return service.toStateFlow()
+  }
+
+  private fun StateFlowZiplineService<T>.toStateFlow(): StateFlow<T> {
+    return object : StateFlow<T> {
+      override suspend fun collect(collector: FlowCollector<T>): Nothing {
+        channelFlow {
```
Comment (colinrtwhite) [inline zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/flows.kt:135]: Using the `Flow.take` operator (and similar operators) fails due to `AbortFlowException` not being caught [here](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/flow/operators/Limit.kt#L67) when used with this `StateFlow` implementation. Wrapping `collect` in a `channelFlow` similar to `FlowSerializer` solves the issue, though I'm not fully sure why.

File: zipline/src/engineTest/kotlin/app/cash/zipline/StateFlowTest.kt [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+    // Confirm that no services or clients were leaked.
+    scope.close()
+    assertEquals(setOf(), endpointA.serviceNames)
+    assertEquals(setOf(), endpointA.clientNames)
+    assertEquals(0, service.mutableFlow.subscriptionCount.first())
+  }
+
+  @Test
+  fun stateFlowAbortWorks() = runBlocking(Unconfined) {
```
Comment (colinrtwhite) [inline zipline/src/engineTest/kotlin/app/cash/zipline/StateFlowTest.kt:120]: This tests for the abort issue mentioned above.

File: zipline/src/engineTest/kotlin/app/cash/zipline/StateFlowTest.kt [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    val scope = ZiplineScope()
+    val (endpointA, endpointB) = newEndpointPair(this)
+    val service = RealStateFlowEchoService(initialValue = "first")
+
+    endpointA.bind<StateFlowEchoService>("service", service)
+    val client = endpointB.take<StateFlowEchoService>("service", scope)
+
+    // 'Flow.take' throws AbortFlowException internally. Make sure we handle it gracefully.
+    assertEquals(listOf("first"), client.flow.take(1).toList())
```
Comment (colinrtwhite) [inline zipline/src/engineTest/kotlin/app/cash/zipline/StateFlowTest.kt:124]: This tests for the abort issue mentioned above.

**Comment by colinrtwhite** [comment]
Hmm seems like one of the tests only fails on the native targets:

```
2023-03-10T07:14:46.6556545Z Cause 1: java.lang.Error: Test running process exited unexpectedly.
2023-03-10T07:14:46.6556992Z Current test: collectFlowOneTime
2023-03-10T07:14:46.6557287Z Process output:
2023-03-10T07:14:46.6557890Z  Uncaught Kotlin exception: kotlinx.coroutines.CompletionHandlerException: Exception in completion handler InvokeOnCompletion@87fda510[job@87feea70] for DeferredCoroutine{Completed}@87feea70
2023-03-10T07:14:46.6558629Z     at 0   test.kexe                           0x527534           kfun:kotlinx.coroutines.JobSupport.notifyCompletion#internal + 660 
2023-03-10T07:14:46.6559301Z     at 1   test.kexe                           0x526bbd           kfun:kotlinx.coroutines.JobSupport.completeStateFinalization#internal + 477 
2023-03-10T07:14:46.6559988Z     at 2   test.kexe                           0x526712           kfun:kotlinx.coroutines.JobSupport.finalizeFinishingState#internal + 2242 
2023-03-10T07:14:46.6560642Z     at 3   test.kexe                           0x52a5cd           kfun:kotlinx.coroutines.JobSupport.tryMakeCompleting#internal + 1693 
2023-03-10T07:14:46.6561252Z     at 4   test.kexe                           0x529d4b           kfun:kotlinx.coroutines.JobSupport#makeCompletingOnce(kotlin.Any?){}kotlin.Any? + 235 
2023-03-10T07:14:46.6561862Z     at 5   test.kexe                           0x5180d6           kfun:kotlinx.coroutines.AbstractCoroutine#resumeWith(kotlin.Result<1:0>){} + 150 
2023-03-10T07:14:46.6562546Z     at 6   test.kexe                           0x420150           kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#resumeWith(kotlin.Result<kotlin.Any?>){} + 464 
2023-03-10T07:14:46.6563632Z     at 7   test.kexe                           0x546bff           kfun:kotlinx.coroutines#resume__at__kotlinx.coroutines.DispatchedTask<0:0>(kotlin.coroutines.Continuation<0:0>;kotlin.Boolean){0§<kotlin.Any?>} + 383 
2023-03-10T07:14:46.6564427Z     at 8   test.kexe                           0x51bfd3           kfun:kotlinx.coroutines.CancellableContinuationImpl.dispatchResume#internal + 755 
2023-03-10T07:14:46.6565138Z     at 9   test.kexe                           0x51c3ff           kfun:kotlinx.coroutines.CancellableContinuationImpl.resumeImpl#internal + 399 
2023-03-10T07:14:46.6565840Z     at 10  test.kexe                           0x51b637           kfun:kotlinx.coroutines.CancellableContinuationImpl#resumeWith(kotlin.Result<1:0>){} + 391 
2023-03-10T07:14:46.6566501Z     at 11  test.kexe                           0x52ccb9           kfun:kotlinx.coroutines.ResumeAwaitOnCompletion.invoke#internal + 329 
2023-03-10T07:14:46.6567180Z     at 12  test.kexe                           0x526b7c           kfun:kotlinx.coroutines.JobSupport.completeStateFinalization#internal + 412 
2023-03-10T07:14:46.6567821Z     at 13  test.kexe                           0x52a3bf           kfun:kotlinx.coroutines.JobSupport.tryMakeCompleting#internal + 1167 
2023-03-10T07:14:46.6568445Z     at 14  test.kexe                           0x529bfb           kfun:kotlinx.coroutines.JobSupport#makeCompleting(kotlin.Any?){}kotlin.Boolean + 171 
2023-03-10T07:14:46.6569085Z     at 15  test.kexe                           0x544a4c           kfun:kotlinx.coroutines.flow.$launchSharingDeferred$lambda$1$lambda$0$FUNCTION_REFERENCE$144.emit#internal + 492 
2023-03-10T07:14:46.6569724Z     at 16  test.kexe                           0x550db6           kfun:kotlinx.coroutines.flow.internal.SafeCollector#emit(1:0){} + 470 
2023-03-10T07:14:46.6570428Z     at 17  test.kexe                           0x5b93b4           kfun:app.cash.zipline.StateFlowTest.RealStateFlowEchoService.$createFlow$lambda$0COROUTINE$61.invokeSuspend#internal + 756 
2023-03-10T07:14:46.6571291Z     at 18  test.kexe                           0x420070           kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#resumeWith(kotlin.Result<kotlin.Any?>){} + 240 
2023-03-10T07:14:46.6571931Z     at 19  test.kexe                           0x5466dc           kfun:kotlinx.coroutines.DispatchedTask#run(){} + 604 
2023-03-10T07:14:46.6572775Z     at 20  test.kexe                           0x54fe9e           kfun:kotlinx.coroutines.WorkerDispatcher.$dispatch$lambda$0$FUNCTION_REFERENCE$151.$<bridge-UNN>invoke(){}#internal + 62 
2023-03-10T07:14:46.6573445Z     at 21  test.kexe                           0x5fd7b1           _ZN6Worker19processQueueElementEb + 1569 
2023-03-10T07:14:46.6573897Z     at 22  test.kexe                           0x5fd11c           _ZN12_GLOBAL__N_113workerRoutineEPv + 92 
2023-03-10T07:14:46.6574307Z     at 23  libc.so.6                           0x7f90185f5b42     0x0 + 140256860920642 
2023-03-10T07:14:46.6574680Z     at 24  libc.so.6                           0x7f90186879ff     0x0 + 140256861518335 
2023-03-10T07:14:46.6575271Z Caused by: app.cash.zipline.ZiplineException: kotlin.UninitializedPropertyAccessException: lateinit property continuation has not been initialized
2023-03-10T07:14:46.6576054Z     at 0   test.kexe                           0x433249           kfun:kotlin.native.internal#ThrowUninitializedPropertyAccessException(kotlin.String){}kotlin.Nothing + 393 
2023-03-10T07:14:46.6577082Z     at 1   test.kexe                           0x556ed4           kfun:app.cash.zipline.internal.bridge.OutboundCallHandler.RealSuspendCallback.<get-continuation>#internal + 148 
2023-03-10T07:14:46.6577991Z     at 2   test.kexe                           0x5571ea           kfun:app.cash.zipline.internal.bridge.OutboundCallHandler.RealSuspendCallback.call#internal + 394 
2023-03-10T07:14:46.6578953Z     at 3   test.kexe                           0x557d8f           kfun:app.cash.zipline.internal.bridge.SuspendCallback.Companion.Adapter.ZiplineFunction0.call#internal + 191 
2023-03-10T07:14:46.6580069Z     at 4   test.kexe                           0x554599           kfun:app.cash.zipline.internal.bridge.Endpoint.object-1.call#internal + 2249 
2023-03-10T07:14:46.6580892Z     at 5   test.kexe                           0x55620c           kfun:app.cash.zipline.internal.bridge.OutboundCallHandler#call(app.cash.zipline.ZiplineService;kotlin.Int;kotlin.Array<out|kotlin.Any?>...){}kotlin.Any? + 1196 
2023-03-10T07:14:46.6581950Z     at 6   test.kexe                           0x558527           kfun:app.cash.zipline.internal.bridge.SuspendCallback.Companion.Adapter.GeneratedOutboundService.success#internal + 247 
2023-03-10T07:14:46.6583218Z     at 7   test.kexe                           0x555b4e           kfun:app.cash.zipline.internal.bridge.InboundService.$callSuspending$lambda$2$FUNCTION_REFERENCE$126.$<bridge-UNNN>invoke(kotlin.Throwable?){}#internal + 910 
2023-03-10T07:14:46.6583970Z     at 8   test.kexe                           0x52ca15           kfun:kotlinx.coroutines.InvokeOnCompletion.invoke#internal + 149 
2023-03-10T07:14:46.6584566Z     at 9   test.kexe                           0x5275c9           kfun:kotlinx.coroutines.JobSupport.notifyCompletion#internal + 809 
2023-03-10T07:14:46.6585228Z     at 10  test.kexe                           0x526bbd           kfun:kotlinx.coroutines.JobSupport.completeStateFinalization#internal + 477 
2023-03-10T07:14:46.6585913Z     at 11  test.kexe                           0x526712           kfun:kotlinx.coroutines.JobSupport.finalizeFinishingState#internal + 2242 
2023-03-10T07:14:46.6586569Z     at 12  test.kexe                           0x52a5cd           kfun:kotlinx.coroutines.JobSupport.tryMakeCompleting#internal + 1693 
2023-03-10T07:14:46.6587171Z     at 13  test.kexe                           0x529d4b           kfun:kotlinx.coroutines.JobSupport#makeCompletingOnce(kotlin.Any?){}kotlin.Any? + 235 
2023-03-10T07:14:46.6587857Z     at 14  test.kexe                           0x5180d6           kfun:kotlinx.coroutines.AbstractCoroutine#resumeWith(kotlin.Result<1:0>){} + 150 
2023-03-10T07:14:46.6588548Z     at 15  test.kexe                           0x420150           kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#resumeWith(kotlin.Result<kotlin.Any?>){} + 464 
2023-03-10T07:14:46.6589582Z     at 16  test.kexe                           0x546bff           kfun:kotlinx.coroutines#resume__at__kotlinx.coroutines.DispatchedTask<0:0>(kotlin.coroutines.Continuation<0:0>;kotlin.Boolean){0§<kotlin.Any?>} + 383 
2023-03-10T07:14:46.6590442Z     at 17  test.kexe                           0x51bfd3           kfun:kotlinx.coroutines.CancellableContinuationImpl.dispatchResume#internal + 755 
2023-03-10T07:14:46.6591158Z     at 18  test.kexe                           0x51c3ff           kfun:kotlinx.coroutines.CancellableContinuationImpl.resumeImpl#internal + 399 
2023-03-10T07:14:46.6591853Z     at 19  test.kexe                           0x51b637           kfun:kotlinx.coroutines.CancellableContinuationImpl#resumeWith(kotlin.Result<1:0>){} + 391 
2023-03-10T07:14:46.6592517Z     at 20  test.kexe                           0x52ccb9           kfun:kotlinx.coroutines.ResumeAwaitOnCompletion.invoke#internal + 329 
2023-03-10T07:14:46.6593188Z     at 21  test.kexe                           0x526b7c           kfun:kotlinx.coroutines.JobSupport.completeStateFinalization#internal + 412 
2023-03-10T07:14:46.6593829Z     at 22  test.kexe                           0x52a3bf           kfun:kotlinx.coroutines.JobSupport.tryMakeCompleting#internal + 1167 
2023-03-10T07:14:46.6594456Z     at 23  test.kexe                           0x529bfb           kfun:kotlinx.coroutines.JobSupport#makeCompleting(kotlin.Any?){}kotlin.Boolean + 171 
2023-03-10T07:14:46.6595104Z     at 24  test.kexe                           0x544a4c           kfun:kotlinx.coroutines.flow.$launchSharingDeferred$lambda$1$lambda$0$FUNCTION_REFERENCE$144.emit#internal + 492 
2023-03-10T07:14:46.6595747Z     at 25  test.kexe                           0x550db6           kfun:kotlinx.coroutines.flow.internal.SafeCollector#emit(1:0){} + 470 
2023-03-10T07:14:46.6596452Z     at 26  test.kexe                           0x5b93b4           kfun:app.cash.zipline.StateFlowTest.RealStateFlowEchoService.$createFlow$lambda$0COROUTINE$61.invokeSuspend#internal + 756 
2023-03-10T07:14:46.6597247Z     at 27  test.kexe                           0x420070           kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#resumeWith(kotlin.Result<kotlin.Any?>){} + 240 
2023-03-10T07:14:46.6597890Z     at 28  test.kexe                           0x5466dc           kfun:kotlinx.coroutines.DispatchedTask#run(){} + 604 
2023-03-10T07:14:46.6598727Z     at 29  test.kexe                           0x54fe9e           kfun:kotlinx.coroutines.WorkerDispatcher.$dispatch$lambda$0$FUNCTION_REFERENCE$151.$<bridge-UNN>invoke(){}#internal + 62 
2023-03-10T07:14:46.6599339Z     at 30  test.kexe                           0x5fd7b1           _ZN6Worker19processQueueElementEb + 1569 
2023-03-10T07:14:46.6599792Z     at 31  test.kexe                           0x5fd11c           _ZN12_GLOBAL__N_113workerRoutineEPv + 92 
2023-03-10T07:14:46.6600204Z     at 32  libc.so.6                           0x7f90185f5b42     0x0 + 140256860920642 
2023-03-10T07:14:46.6600574Z     at 33  libc.so.6                           0x7f90186879ff     0x0 + 140256861518335 
2023-03-10T07:14:46.6600782Z 
2023-03-10T07:14:46.6601178Z     at 0   test.kexe                           0x5614a8           kfun:app.cash.zipline.internal.bridge.ThrowableSerializer.$deserialize$lambda$0$FUNCTION_REFERENCE$137.invoke#internal + 104 
2023-03-10T07:14:46.6602015Z     at 1   test.kexe                           0x5613cb           kfun:app.cash.zipline.internal.bridge.ThrowableSerializer#deserialize(kotlinx.serialization.encoding.Decoder){}kotlin.Throwable + 1467 
2023-03-10T07:14:46.6603310Z     at 2   test.kexe                           0x4fffce           kfun:kotlinx.serialization.json.internal.StreamingJsonDecoder#decodeSerializableValue(kotlinx.serialization.DeserializationStrategy<0:0>){0§<kotlin.Any?>}0:0 + 238 
2023-03-10T07:14:46.6604894Z     at 3   test.kexe                           0x500aaf           kfun:kotlinx.serialization.json.internal.StreamingJsonDecoder#decodeSerializableElement(kotlinx.serialization.descriptors.SerialDescriptor;kotlin.Int;kotlinx.serialization.DeserializationStrategy<0:0>;0:0?){0§<kotlin.Any?>}0:0 + 1039 
2023-03-10T07:14:46.6606167Z     at 4   test.kexe                           0x55a768           kfun:app.cash.zipline.internal.bridge.ResultOrCallbackSerializer#deserialize(kotlinx.serialization.encoding.Decoder){}app.cash.zipline.internal.bridge.ResultOrCallback<1:0> + 1384 
2023-03-10T07:14:46.6607609Z     at 5   test.kexe                           0x4fffce           kfun:kotlinx.serialization.json.internal.StreamingJsonDecoder#decodeSerializableValue(kotlinx.serialization.DeserializationStrategy<0:0>){0§<kotlin.Any?>}0:0 + 238 
2023-03-10T07:14:46.6608805Z     at 6   test.kexe                           0x4efe81           kfun:kotlinx.serialization.json.Json#decodeFromString(kotlinx.serialization.DeserializationStrategy<0:0>;kotlin.String){0§<kotlin.Any?>}0:0 + 897 
2023-03-10T07:14:46.6609637Z     at 7   test.kexe                           0x5577d7           kfun:app.cash.zipline.internal.bridge.OutboundCallHandler.$call$lambda$0$FUNCTION_REFERENCE$128.invoke#internal + 247 
2023-03-10T07:14:46.6610675Z     at 8   test.kexe                           0x5532ba           kfun:app.cash.zipline.internal.bridge.Endpoint#withTakeScope(app.cash.zipline.ZiplineScope;kotlin.Function0<0:0>){0§<kotlin.Any?>}0:0 + 186 
2023-03-10T07:14:46.6611569Z     at 9   test.kexe                           0x556278           kfun:app.cash.zipline.internal.bridge.OutboundCallHandler#call(app.cash.zipline.ZiplineService;kotlin.Int;kotlin.Array<out|kotlin.Any?>...){}kotlin.Any? + 1304 
2023-03-10T07:14:46.6612621Z     at 10  test.kexe                           0x558527           kfun:app.cash.zipline.internal.bridge.SuspendCallback.Companion.Adapter.GeneratedOutboundService.success#internal + 247 
2023-03-10T07:14:46.6613886Z     at 11  test.kexe                           0x555b4e           kfun:app.cash.zipline.internal.bridge.InboundService.$callSuspending$lambda$2$FUNCTION_REFERENCE$126.$<bridge-UNNN>invoke(kotlin.Throwable?){}#internal + 910 
2023-03-10T07:14:46.6614613Z     at 12  test.kexe                           0x52ca15           kfun:kotlinx.coroutines.InvokeOnCompletion.invoke#internal + 149 
2023-03-10T07:14:46.6615235Z     at 13  test.kexe                           0x5275c9           kfun:kotlinx.coroutines.JobSupport.notifyCompletion#internal + 809 
```

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/flows.kt [hunk 4]
```diff
@@ -86,3 +91,65 @@ internal class FlowSerializer<T>(
...
+          this@toStateFlow.collect(object : FlowZiplineCollector<T> {
+            override suspend fun emit(value: T) {
+              this@channelFlow.send(value)
+            }
+          })
+        }.collect(collector)
+        throw AssertionError() // StateFlows never complete!
```
Comment (swankjesse) [inline zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/flows.kt:142]: nice comment

**Review by swankjesse — APPROVED** [review]

**Comment by swankjesse** [comment]
I’m happy to investigate the native crash.

**Comment by colinrtwhite** [comment]
@swankjesse That'd be awesome, thanks! I'm suspicious it could be another compiler bug because the coroutine completes successfully before `suspendCancellableCoroutine`'s block is invoked.

**Comment by colinrtwhite** [comment]
~Made some progress on this. The tests are failing due to a race condition somewhere that causes `Endpoint.bind` to be called before `Endpoint.remove`. As a result the `collectFlowZeroTimes` sometimes fails as `endpointA.serviceNames` isn't empty. I updated `Endpoint` to keep track of removed services, which "fixes" the tests, but isn't a mergeable solution.~

EDIT: I think the race condition might have been in the tests? I removed the code that changed dispatchers and the tests now pass.

**Review by JakeWharton — APPROVED** [review]

**Comment by swankjesse** [comment]
# amazing! 🥳

**Pull request merged**